numba/testing.py

```diff
@@ -1,9 +1,522 @@
 from __future__ import print_function, division, absolute_import
 
 import sys
+import os
 import functools
-
+import numba.unittest_support as unittest
+from unittest import result, runner, signals, suite, loader, case
+from numba.utils import PYVERSION, StringIO
 from numba import config
+import argparse
+import collections
+import contextlib
+import cProfile
+import multiprocessing
+from fnmatch import fnmatch
+from os.path import dirname, join, abspath, relpath
+
+try:
+    from multiprocessing import TimeoutError
+except ImportError:
+    from Queue import Empty as TimeoutError
+
+
+class TestLister(object):
+    """Simply list available tests rather than running them."""
+    def __init__(self):
+        pass
+
+    def run(self, test):
+        result = runner.TextTestResult(sys.stderr, True, 1)
+        self._test_list = _flatten_suite(test)
+        for t in self._test_list:
+            print(t.id())
+        print('%d tests found'%len(self._test_list))
+        return result
+        
+class SerialSuite(unittest.TestSuite):
+    """A simple marker to make sure tests in this suite are run serially."""
+    
+    pass
+
+def load_testsuite(loader, dir):
+    """Find tests in 'dir'."""
+    
+    suite = unittest.TestSuite()
+    files=[]
+    for f in os.listdir(dir):
+        path=join(dir, f)
+        if os.path.isfile(path) and fnmatch(f, 'test_*.py'):
+            files.append(f)
+        elif os.path.isfile(join(path, '__init__.py')):
+            suite.addTests(loader.discover(path))
+    for f in files:
+        # turn 'f' into a filename relative to the toplevel dir...
+        f = relpath(join(dir, f), loader._top_level_dir)
+        # ...and translate it to a module name.
+        f = os.path.splitext(os.path.normpath(f.replace(os.path.sep, '.')))[0]
+        suite.addTests(loader.loadTestsFromName(f))
+    return suite
+
+class TestLoader(loader.TestLoader):
+
+    _top_level_dir=dirname(dirname(__file__))
+    
+    def discover(self, start_dir, pattern='test*.py', top_level_dir=None):
+        """Upstream discover doesn't consider top-level 'load_tests' functions.
+        If we find load_tests in start_dir, deal with it here. Otherwise
+        forward the call to the base class."""
+        top = top_level_dir or self._top_level_dir
+        location = top and join(abspath(top), start_dir) or start_dir
+        if os.path.isfile(join(location, '__init__.py')):
+            name = self._get_name_from_path(location)
+            try:
+                package = self._get_module_from_name(name)
+                load_tests = getattr(package, 'load_tests', None)
+                if load_tests:
+                    return load_tests(self, [], pattern)
+            except case.SkipTest as e:
+                return _make_skipped_test(name, e, self.suiteClass)
+            else:
+                return super(TestLoader, self).discover(start_dir, pattern, top)
+        else:
+            try:
+                __import__(start_dir)
```

Comment:
Hopefully not. But I have to admit that I struggled quite a bit trying to figure out the possible values of the inputs. (The docs are very sparse, and the code shows its semantic evolution, so it isn't quite clear what those variables are supposed (and allowed) to contain.)